src/ore/src/assert.rs

```diff
@@ -14,6 +14,97 @@
 // limitations under the License.
 
 //! Assertion utilities.
+//!
+//! # Soft assertions
+//!
+//! Soft assertions are like debug assertions, but they can be toggled on and
+//! off at runtime in a release build.
+//!
+//! They are useful in two scenarios:
+//!
+//!   * When a failed assertion should result in a log message rather a process
+//!     crash.
+//!
+//!   * When evaluating the condition is too expensive to evaluate in production
+//!     deployments.
+//!
+//! When soft assertions are disabled, the performance cost of each assertion is
+//! one branch on an atomic.
+//!
+//! Ore provides the following macros to make soft assertions:
+//!
+//!   * [`soft_assert`](crate::soft_assert)
+//!   * [`soft_assert_eq`](crate::soft_assert_eq)
+//!   * [`soft_assert_or_log`](crate::soft_assert_or_log)
+//!   * [`soft_panic_or_log`](crate::soft_panic_or_log)
+//!
+//! Due to limitations in Rust, these macros are exported at the crate root.
+
+use std::sync::atomic::AtomicBool;
+
+use crate::env;
+
+/// Whether to enable soft assertions.
+///
+/// This value defaults to `true` when `debug_assertions` are enabled, or to the
+/// value of the `MZ_SOFT_ASSERTIONS` environment variable otherwise.
+#[ctor::ctor]
```

Comment:
Can we make this an optional dependency please?

Also, is this even an allowed usage of this crate? The docs for it say you can't use the rust "stdlib services" in ctor functions. Dunno what "services" means here but env is of course stdlib. From https://crates.io/crates/ctor/0.1.20:

> Rust's philosophy is that nothing happens before or after main and this library explicitly subverts that. The code that runs in the ctor and dtor functions should be careful to limit itself to libc functions and code that does not rely on Rust's stdlib services.

Reply:
I trawled through the code in `env` last week and it didn't seem like it was using anything that doesn't ultimately come from libc.

Happy to have another set of eyes on it though.

Reply:
@benesch said something like "if it'd make you happier, I could just switch this to call `getenv` from libc".

Maybe that's still worth doing? Even though I'm pretty sure our use of `std` here is safe, it'd be a lot clearer to readers of the code if we just had one simple libc call.

Reply:
That's sufficient due diligence for me. We should add a comment to that effect here though, to prevent future confusion.

Reply:
Done.